### PR TITLE
fixes #14511 [backport:1.4]

### LINKED ISSUE #14511: Can't attach a pragma to a cased field.
Attaching a pragma to a field used in a case within an object has no effect on the AST.

### Example
```nim
import macros

template myPragma() {.pragma.}

type
  XType = enum
    XInt,
    XString
  X = object
    case y {.myPragma.}: XType
      of XInt:
        a: int
      else:
        b: string
  
var x: X = X(y: XInt, a: 5)
echo x.y.hasCustomPragma(myPragma)
echo x
echo X(y: XString, b: "abc")
```

### Current Output
```
nim/lib/core/macros.nim(1546, 25) Error: index 1 not in 0 .. 0
```

### Expected Output
```
true
(y: XInt, a: 5)
(y: XString, b: "abc")
```

### Additional Information
I did also try `X.y.hasCustomPragma(myPragma)`; that compiles, yet returns false.

```
$ nim -v
Nim Compiler Version 0.1.2
```

### PR DESCRIPTION
**Pull request by Dankr4d**
fixes #14511

### CODE REVIEW

File: lib/core/macros.nim [hunk 1]
```diff
@@ -1545,7 +1545,12 @@ proc customPragmaNode(n: NimNode): NimNode =
               let varNode = identDefs[i]
               # if it is and empty branch, skip
               if varNode[0].kind == nnkNilLit: continue
-              if varNode[1].kind == nnkIdentDefs:
+              # Skiping, because required ident node is only available in node of kind nnkOfBranch
```
Comment (konsumlamm) [inline lib/core/macros.nim]:
```suggestion
              # Skipping, because required ident node is only available in node of kind nnkOfBranch
```

Comment (Dankr4d) [reply lib/core/macros.nim]: I force pushed my changes, since the logic I implemented here wasn't correct.

File: lib/core/macros.nim [hunk 2]
```diff
@@ -1545,11 +1545,15 @@ proc customPragmaNode(n: NimNode): NimNode =
...
+              if varNode.len > 1 and varNode[1].kind == nnkIdentDefs:
                 identDefsStack.add(varNode[1])
               else: # nnkRecList
-                for j in 0 ..< varNode[1].len:
-                  identDefsStack.add(varNode[1][j])
+                var idx: int = 1
+                if varNode.kind == nnkElse:
+                  idx = 0
+                for j in 0 ..< varNode[idx].len:
+                  identDefsStack.add(varNode[idx][j])
```
Comment (alaviss) [inline lib/core/macros.nim]: I think it's better to rewrite this piece using a case statement to dispatch between node types and move nnkRecList handling outside since the current version is too hard to follow, something like this perhaps:

```nim
case identDefs.kind
of nnkRecList:
  for child in identDefs.items:
    identDefsStack.add(child)
of nnkRecCase:
  # Add condition definition
  identDefsStack.add(identDefs[0])
  # Add branches
  for i in 1 ..< identDefs.len:
    let varNode = identDefs[i]
    case varNode.kind
    of nnkOfBranch:
      # We are only interested in variable definitions
      if varNode[1].kind in {nnkIdentDefs, nnkRecList}:
        identDefsStack.add(varNode[1])
    of nnkElse:
      # We are only interested in variable definitions
      if varNode[0].kind in {nnkIdentDefs, nnkRecList}:
        identDefsStack.add(varNode[0])
    else:
      doAssert false, "Unknown node kind in nnkRecCase: " & $varNode.kind
else:
  <insert identdefs handling code here>
```

Though the best thing to do would be to remove this mess that is `identDefsStack` and use a recursive function, but that can be left for later.

Comment (Dankr4d) [reply lib/core/macros.nim]: I've replaced the fix with your cleaner version, with small changes.
I'm calling the `children` instead of the `items` iterator and added brackets to the `doAssert` call. I think it's easier to read.

File: lib/core/macros.nim [hunk 3]
```diff
@@ -1539,18 +1539,28 @@ proc customPragmaNode(n: NimNode): NimNode =
...
-              if varNode[1].kind == nnkIdentDefs:
-                identDefsStack.add(varNode[1])
-              else: # nnkRecList
-                for j in 0 ..< varNode[1].len:
-                  identDefsStack.add(varNode[1][j])
-
+              case varNode.kind
+              of nnkOfBranch:
+                # We are only interested in variable definitions
+                if varNode[1].kind in {nnkIdentDefs, nnkRecList}:
```
Comment (Araq) [inline lib/core/macros.nim]: You need to use the last son of `varNode`, not `[1]` here.

File: tests/macros/t14511.nim [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  template myPragma() {.pragma.}
+
+  type
+    XType = enum
+      XInt,
+      XString
+    X = object
+      case y {.myPragma.}: XType
+        of XInt:
```
Comment (Araq) [inline tests/macros/t14511.nim]: Make this `of XInt, XUnused` to see that your version still didn't work properly.

File: lib/core/macros.nim (hunk 3 above)
Comment (alaviss) [reply lib/core/macros.nim]: Then the dispatch can be removed and replaced with `varNode.last`

Comment (Dankr4d) [reply lib/core/macros.nim]: I've replaced the dispatch code, like @alaviss mentioned. The last node is always of kind nnkRecList so it's processed the next iteration.

File: tests/macros/t14511.nim (hunk 4 above)
Comment (Dankr4d) [reply tests/macros/t14511.nim]: What exactly do you mean?
When replacing XInt with XUnused or adding it to (like above), then all tests are still successful. I also cannot see the issue here. It only crashes when in object initialization the wrong XType is passed.

File: tests/macros/t14511.nim [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+block TEST_1:
+  # https://github.com/nim-lang/Nim/issues/14511
+
+  template myPragma() {.pragma.}
+
+  type
+    XType = enum
+      XInt,
+      XString
```
Comment (Araq) [inline tests/macros/t14511.nim]:
```suggestion
      XString,
      XUnused
```

File: tests/macros/t14511.nim (hunk 4 above)
Comment (Araq) [reply tests/macros/t14511.nim]:
```suggestion
        of XInt, XUnused:
```
